packages/playwright/src/transform/transform.ts

```diff
@@ -81,12 +83,44 @@ function validateTsConfig(tsconfig: LoadedTsConfig): ParsedTsConfigData {
 }
 
 function loadAndValidateTsconfigsForFile(file: string): ParsedTsConfigData[] {
-  const tsconfigPathOrDirecotry = _singleTSConfig || path.dirname(file);
-  if (!cachedTSConfigs.has(tsconfigPathOrDirecotry)) {
-    const loaded = tsConfigLoader(tsconfigPathOrDirecotry);
-    cachedTSConfigs.set(tsconfigPathOrDirecotry, loaded.map(validateTsConfig));
+  if (_singleTSConfigPath && !_singleTSConfig)
+    _singleTSConfig = loadTsConfig(_singleTSConfigPath).map(validateTsConfig);
+  if (_singleTSConfig)
+    return _singleTSConfig;
+  return loadAndValidateTsconfigsForFolder(path.dirname(file));
+}
+
+function loadAndValidateTsconfigsForFolder(folder: string): ParsedTsConfigData[] {
+  const foldersWithConfig: string[] = [];
+  let currentFolder = folder;
+  let result: ParsedTsConfigData[] = [];
+  while (true) {
+    const cached = cachedTSConfigs.get(currentFolder);
+    if (cached) {
+      result = cached;
+      break;
+    }
+
+    foldersWithConfig.push(currentFolder);
+
+    for (const name of ['tsconfig.json', 'jsconfig.json']) {
+      const configPath = path.join(currentFolder, name);
+      if (fs.existsSync(configPath)) {
+        const loaded = loadTsConfig(configPath);
+        result = loaded.map(validateTsConfig);
+        break;
+      }
+    }
+
+    const parentFolder = path.resolve(currentFolder, '../');
+    if (currentFolder === parentFolder)
+      break;
+    currentFolder = parentFolder;
   }
-  return cachedTSConfigs.get(tsconfigPathOrDirecotry)!;
+
+  for (const folder of foldersWithConfig)
+    cachedTSConfigs.set(folder, result);
```

Comment:
Maybe store only actual folder with the config since we anyway check the cache for each parent folder now?

Reply:
that would save us the walk up the folder tree